File: Controllers/SubDepartmentController.js
const express = require("express");
const router = express.Router();
// const { SubDepartment } = require("../models");
const db = require('../config/database'); 
const SubDepartment = db.SubDepartment
const { Op } = require("sequelize");

const requireAuth=require("../middleware/requireAuth")

// Route: Index (List)
router.get("/", requireAuth, async (req, res) => {
    console.log("req===>",req.user)
  const username = req.user.userName;
  // Only list active subdepartments and include parent department (if any)
  const list = await SubDepartment.findAll({
    where: { Active: 1 },
    include: [
      { model: db.Department, required: false }
    ]
  });
  res.json({ list, username });
});
// router.get("/", requireAuth, async (req, res) => {
//   console.log("req===>", req.user);
//   const username = req.user.userName;

//   const list = await SubDepartment.findAll({
//     include: [
//       {
//         model: db.Department,
//         required: true // ensures only SubDepartments with a Department are included
//       }
//     ]
//   });

//   res.json({ list, username });
// });


// Route: Create Form
router.get("/create", requireAuth, (req, res) => {
  const username = req.session.username;
  res.render("subDepartment/create", { username });
});

// Route: Handle Create
router.post("/create", requireAuth, async (req, res) => {
  const { Name, Code,DepartmentID } = req.body;
  const username = req.user.username;

  const nameExist = await SubDepartment.findOne({ where: { Name } });
  const codeExist = await SubDepartment.findOne({ where: { Code } });

  if (!Name || nameExist) {
    return res.json( { NameAlert: "Name already exists or is null" });
  }

  if (!Code || codeExist) {
    return res.json( { CodeAlert: "Code already exists or is null" });
  }

  await SubDepartment.create({
    Name,
    DepartmentID:DepartmentID,
    Code,
    CreatedBy: username,
    CreatedDate: new Date(),
    Active: true,
  });

  res.json({
    status:true
  });
});

// Route: Edit Form
router.get("/edit/:id", requireAuth, async (req, res) => {
  // Only allow editing active subdepartments
  const subDept = await SubDepartment.findOne({ where: { id: req.params.id, Active: 1 } });
  if (!subDept) return res.status(404).send("Not found");
  req.session.subDeptBackup = subDept; // Store original
  res.render("subDepartment/edit", { subDept });
});

// Route: Handle Edit
router.post("/edit/:id", requireAuth, async (req, res) => {
  const { Name, Code,DepartmentID } = req.body;
  const id = req.params.id;
  const username = req.user.userName;
 

  const nameExist = await SubDepartment.findOne({ where: { Name, id: { [Op.ne]: id } } });
  const codeExist = await SubDepartment.findOne({ where: { Code, id: { [Op.ne]: id } } });

  if (!Name || nameExist) {
    return res.json( { NameAlert: "Name already exists or is null", subDept: req.body });
  }

  if (!Code || codeExist) {
    return res.json( { CodeAlert: "Code already exists or is null", subDept: req.body });
  }

  await SubDepartment.update(
    {
      Name,
      Code,
      ModifyBy: username,
      DepartmentID:DepartmentID,
      ModifyDate: new Date(),
    },
    { where: { id } }
  );

  res.json({
    status:true
  });
});

// Route: Delete
router.get("/delete/:id", requireAuth, async (req, res) => {
  // Soft-delete: set Active = 0 instead of destroying the record
  await SubDepartment.update(
    { Active: 0 },
    { where: { id: req.params.id } }
  );

  res.json({ status: true });
});

// Route: Search
router.post("/search", requireAuth, async (req, res) => {
  const inputValue = req.body.inputValue;
  if (!inputValue) return res.redirect("/subdepartments");

  const nameMatches = await SubDepartment.findAll({
    where: { Active: 1, Name: { [Op.like]: `%${inputValue}%` } },
  });

  const codeMatches = await SubDepartment.findAll({
    where: { Active: 1, Code: { [Op.like]: `%${inputValue}%` } },
  });

  // Merge unique
  const allMatches = [...new Map([...nameMatches, ...codeMatches].map(item => [item.id, item])).values()];
  res.json({ list: allMatches });
});

module.exports = router;

File: Controllers/SubDepartmentController.cs
using DMS.Models;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using System.Diagnostics;

namespace DMS.Controllers
{
    public class SubDepartmentController : Controller
    {
        public const string SessionKeyName = "username";
        private readonly ILogger<SubDepartmentController> _logger;
        private readonly IHttpContextAccessor contxt;

        private DMS_DbContext dataContext { get; set; }

        public SubDepartmentController(ILogger<SubDepartmentController> logger, DMS_DbContext _dataContext, IHttpContextAccessor httpContextAccessor)
        {
            _logger = logger;
            dataContext = _dataContext;
            contxt = httpContextAccessor;
        }

        public IActionResult Create()
        {
            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View();
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Create(SubDepartmentModel subDepartment)
        {
            // Validate if the name and age exist in the database
            bool nameExist = dataContext.SubDepartment.Any(p => p.Name == subDepartment.Name);
            bool codeExist = dataContext.SubDepartment.Any(p => p.Code == subDepartment.Code);

            if (nameExist)
            {
                ViewBag.NameAlert = "<span class='text-danger'>This Name already Exist</span>";
                return View(subDepartment);
            }
            if (subDepartment.Name == null)
            {
                ViewBag.NameAlert = "<span class='text-danger'>Please Input a Name</span>";
                return View(subDepartment);
            }
            if (codeExist)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>This Code already Exist</span>";
                return View(subDepartment);
            }
            if (subDepartment.Code == null)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>Please Input a Code</span>";
                return View(subDepartment);
            }

            subDepartment.CreatedDate = DateTime.Now;
            subDepartment.Active = true;
            subDepartment.CreatedBy = HttpContext.Session.GetString(SessionKeyName);
            dataContext.SubDepartment.Add(subDepartment);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Edit(SubDepartmentModel subDepartment)
        {
            var name = (string)TempData["Name"];
            var code = (string)TempData["Code"];

            // Validate if the name and age exist in the database
            bool nameExist = dataContext.SubDepartment.Any(p => p.Name == subDepartment.Name && p.Name != name);
            bool codeExist = dataContext.SubDepartment.Any(p => p.Code == subDepartment.Code && p.Code != code);

            var active = (bool)TempData["Active"];
            var CreatedBy = (string)TempData["CreatedBy"];
            var CreatedDate = (DateTime)TempData["CreatedDate"];

            if (subDepartment.Name == name && subDepartment.Code == code)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>You don't change any of the data!</span>";
                LoadSubDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(subDepartment);
            }
            if (nameExist)
            {
                ViewBag.NameAlert = "<span class='text-danger'>This Name already Exist</span>";
                LoadSubDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(subDepartment);
            }
            if (subDepartment.Name == null)
            {
                ViewBag.NameAlert = "<span class='text-danger'>Please Input a Name</span>";
                LoadSubDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(subDepartment);
            }
            if (codeExist)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>This Code already Exist</span>";
                LoadSubDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(subDepartment);
            }
            if (subDepartment.Code == null)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>Please Input a Code</span>";
                LoadSubDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(subDepartment);
            }

            subDepartment.CreatedDate = CreatedDate;
            subDepartment.Active = active;
            subDepartment.CreatedBy = CreatedBy;
            subDepartment.ModifyDate = DateTime.Now;
            subDepartment.ModifyBy = HttpContext.Session.GetString(SessionKeyName);
            dataContext.SubDepartment.Update(subDepartment);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        public void LoadSubDepartmentTemp(string name, string code, bool active, string CreatedBy, DateTime CreatedDate)
        {
            TempData["Name"] = name;
            TempData["Code"] = code;
            TempData["Active"] = active;
            TempData["CreatedBy"] = CreatedBy;
            TempData["CreatedDate"] = CreatedDate;
        }

        public IActionResult Edit(long id)
        {
            if (id == null || id == 0)
                return NotFound();

            var subDepartment = dataContext.SubDepartment.Find(id);

            if (subDepartment == null)
                return NotFound();

            TempData["Name"] = subDepartment.Name;
            TempData["Code"] = subDepartment.Code;
            TempData["Active"] = subDepartment.Active;
            TempData["CreatedBy"] = subDepartment.CreatedBy;
            TempData["CreatedDate"] = subDepartment.CreatedDate;
            return View(subDepartment);
        }

        public IActionResult Delete(long id)
        {
            var obj = dataContext.SubDepartment.Find(id);
            if (obj == null)
                return NotFound();

            dataContext.SubDepartment.Remove(obj);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        public IActionResult Index()
        {
            // Retrieve the data from TempData
            var finalListJson = TempData["FinalList"] as string;

            if (finalListJson != null)
            {
                // Deserialize back to List<SubDepartmentModel>
                var finalList = JsonConvert.DeserializeObject<List<SubDepartmentModel>>(finalListJson);

                // Pass it to the view or process it as needed
                return View(finalList);
            }

            IEnumerable<SubDepartmentModel> SubDepartmentList = dataContext.SubDepartment;
            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View(SubDepartmentList);
        }

        [HttpPost]
        public IActionResult Search(string? inputValue)
        {
            if (inputValue == null)
            {
                return RedirectToAction("Index");
            }

            List<SubDepartmentModel> nameSubDepartments = dataContext.SubDepartment
                                                            .Where(d => d.Name.Contains(inputValue))
                                                            .ToList();

            List<SubDepartmentModel> codeSubDepartments = dataContext.SubDepartment
                                                            .Where(d => d.Code.Contains(inputValue))
                                                            .ToList();

            var FinalList = nameSubDepartments.Concat(codeSubDepartments)
                                .Distinct()
                                .ToList();

            // Store in TempData (need to serialize to JSON or similar format)
            TempData["FinalList"] = JsonConvert.SerializeObject(FinalList);

            // Redirect to another action
            return RedirectToAction("Index");
        }

        [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
        public IActionResult Error()
        {
            return View(new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
        }
    }
}
